clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1080,8 +1080,7 @@ static ParamDesc makeParamDesc(ASTContext &Ctx, StringRef Name, QualType Ty) {
 }
 
 static void unsupportedFreeFunctionParamType() {
-  llvm::report_fatal_error("Only scalars and pointers are permitted as "
-                           "free function parameters");
+  llvm::report_fatal_error("Unsupported free kernel parameter type!");
```

Comment:
Is that function still used after your changes? I think we should remove it. Such diagnosing should happen with Sema::Diag, not llvm::report_fatal_error.

Reply:
It is used in a bunch of places. I think this should be done in the scope of another PR as its not directly related to the changes here.